src/physics/utilities/equation_solver.cpp

```diff
@@ -38,6 +38,169 @@ EquationSolver::EquationSolver(MPI_Comm comm, const LinearSolverParameters& lin_
+   */
+  void add(const JSONTable& other)
+  {
+    literals_.insert(other.literals_.begin(), other.literals_.end());
+    for (const auto& [key, subtable] : other.tables_) {
+      operator[](key).add(*subtable);
+    }
+  }
+
+  /**
+   * @brief Inserts the JSON representation to a stream
+   * @param[inout] out The stream to insert into
+   * @param[in] table The table to be inserted (printed)
+   */
+  friend std::ostream& operator<<(std::ostream& out, const JSONTable& table);
+
+private:
+  /**
+   * @brief Literal (primitive) members of the table
+   */
+  JSONLiterals literals_;
+  /**
+   * @brief Subtable (object) members of the table
+   */
+  std::unordered_map<std::string, std::unique_ptr<JSONTable>> tables_;
+  /**
+   * @brief Current depth of the table relative to the root
+   */
+  const int depth_;
+};
+
+std::ostream& operator<<(std::ostream& out, const JSONTable& table)
+{
+  out << "{";
+  std::string indent(table.depth_ * 2, ' ');  // Double-space indenting
+  char        sep = ' ';                      // Start with empty separator to avoid trailing comma
+  for (const auto& [key, val] : table.literals_) {
+    out << sep << "\n" << indent << "\"" << key << "\": ";
+    // Strings need to be wrapped in escaped quotes
+    if (auto str_ptr = std::get_if<std::string>(&val)) {
+      out << "\"" << *str_ptr << "\"";
+    } else {
+      std::visit([&out](const auto contained_val) { out << contained_val; }, val);
+    }
+    sep = ',';
+  }
+
+  // Recursively insert subtables into the stream
+  for (const auto& [key, subtable] : table.tables_) {
+    out << sep << "\n" << indent << "\"" << key << "\": ";
+    out << *subtable;
+    sep = ',';
+  }
+  out << "\n" << indent << "}";
+  return out;
+}
+
+std::unique_ptr<mfem::AmgXSolver> configureAMGX(const MPI_Comm comm, const AMGXPrec& options)
+{
+  auto              amgx                 = std::make_unique<mfem::AmgXSolver>();
+  static const auto default_prec_options = []() {
+    JSONTable result;
+    result.literal("solver")       = "AMG";
+    result.literal("presweeps")    = 1;
+    result.literal("postsweeps")   = 2;
+    result.literal("interpolator") = "D2";
+    result.literal("max_iters")    = 2;
+    result.literal("convergence")  = "ABSOLUTE";
+    result.literal("cycle")        = "V";
+
+    JSONTable top_level;
+    top_level.literal("config_version") = 2;
+    top_level["solver"].add(result);
+    return top_level;
+  }();
+
+  JSONTable options_table;
+  options_table.add(default_prec_options);
+  if (options.verbose) {
+    static const auto default_verbose_options = []() {
+      JSONTable result;
+      result.literal("obtain_timings")    = 1;
+      result.literal("monitor_residual")  = 1;
+      result.literal("print_solve_stats") = 1;
+      result.literal("print_solve_stats") = 1;
+      return result;
+    }();
+    options_table["solver"].add(default_verbose_options);
+  }
+
+  // FIXME: magic_enum?
+  static const auto solver_names = []() {
```

Comment:
I agree this is a bit gross. magic_enum seems like a good way out of it.

Reply:
Unfortunately I don't think this is an option for us because it requires GCC 9+.  There's quite a few instances where we'd benefit from this (mostly other enums in `solver_config.hpp`) so if/when we're able to upgrade our GCC we can look into this.